Demystifying large language models

pm0kjp opened the pull request.

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,691 @@
+
+## Benefits of LLMs
+
+Any process that uses language can potentially benefit from LLM technology.  Consider some of these use cases, which we supplement with examples of prompts we supplied and replies we received from ChatGPT:
```
Consider revamping this page / the prompts to produce much shorter examples? If the intention is for folks to actually engage with the examples, these are so long I don't think anyone will read more than a line or two. This module is already really text heavy, and this page could contribute to a sense of overwhelm by how dense the module is (Even if its not expected for folks to carefully read everything)

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,691 @@
+
+-------
+
+When prompted, **"Create a multiple choice question with several correct answers that asks 'which of the following are true statements about large language models'"**, ChatGPT gave us the following, which we'd like you to answer:   
```
```suggestion
When prompted, **"Create a multiple choice question with several correct answers that asks 'which of the following are true statements about large language models'"**, ChatGPT gave us the following:
```
apparent that we want them to answer

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,691 @@
+>
+>These explanations clarify why each statement is either true or false based on the characteristics and capabilities of large language models.
+
+We checked the questions and answers for accuracy, and we think you'll agree that these are high-quality questions and explanations!
```
```suggestion
```

redundant

franzenr replied:
That's fair! My thought here is that this has been modelled this pretty well throughout the entire module (and perhaps more deftly in some other places, like criticism/feedback to some of the outputs is given, including lower down on this page). As a result, I was thinking this is one place where it felt a bit unnecessary or possibly even tiresome to include again (and also this module is very text heavy, so I was looking to identify some places for reduction). 

But I'm not so stuck on this that I wouldn't pass QA over it or anything like that, so if you think it's important to have in there or that it serves a distinct pedagogical purpose from other ways we're communicating this, then by all means ignore this suggestion!

pm0kjp replied:
OK I think you've convinced me!  Will omit.

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,691 @@
+
+We don't know what your research entails, or what your intentions are for the use of AI in your career.  However, we encourage you to:
+
+* Make no assumptions about the accuracy of any text generated by LLMs 
```
So the question above is about how to improve/accelerate work, but this answer explanation is about what pitfalls to be careful of? 

If you want to include these reminders to drive it home, that's fine, but I think it would make more sense to first also include some examples that address the question that was asked.

pm0kjp replied:
Agree, I added additional context here, @franzenr, tell me what you think!

franzenr replied:
hmmm, I'm not seeing any changes to this section? is it possible the changes didn't get pushed? @pm0kjp

franzenr replied:
jk, that was a me problem. Not sure why I wasn't seeing it, but I can see it now! 🤦

franzenr replied:
LGTM :)

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,691 @@
+
+Because the training data given to LLMs is so vast, there are cases (some involving litigation) of private information such as home addresses being leaked by LLMs.  Additionally, LLMs might correctly infer and disclose sensitive information while amassing and summarizing data (for example, accurately disclosing a serious disease diagnosis to an unprepared user asking about their symptoms).
+
+While the authors do not explicitly mention this, another information hazard we want to point out is the risk related to input text.  Users can disclose private or regulated information to an LLM via prompts.  One of the ways that LLMs acquire additional language samples is by adding the prompts that users supply to their vast collection of language examples.  Supplying private information (about ...
```
This doesn't match my understanding of what an information hazard is, at least as originally defined by Bostrom. An information hazard is the hazard that arises from someone having access to information that can allow them to cause harm to someone else, or causes harm to them directly (e.g., harm of finding out you have a disease while being unprepared, not necessarily having access to supports, etc)  

This example seems to moreso be something that is upstream of an information hazard -- that is, it's an action that could add to model training in a way that _creates future information hazards_, but is not an information hazard itself.

pm0kjp replied:
I'm taking this directly from the cited article and accurately represents Weidinger et al., so let's leave it be, and allow the primary authors to duke terminology out amongst themselves.

franzenr replied:
Oops, looks like my comment may have been on too big of a section of the module -- I'm referring specifically only to the paragraph that starts with "While the authors do not explicitly mention this," 

If it is the case that you think this is in line with what the authors are saying, it might be worth rewriting the start of the sentence then, as I took this as your own riff on things, and not that the authors of the article called providing private information as input text an information hazard itself. (This is in line with my own reading of that section of the article, but I didn't read the entire article itself). 

I also have a lot of friends working in the AI safety space (of which Bostrom's "Superintelligence" is a seminal text) who do a lot of thinking + talking about info hazards -- could easily get a quick look over from them (by end of day) to see if they think this is the right place to frame this.

pm0kjp replied:
OH OH OH gotcha you're right, can you make a stab at a suggestion for a change?  Maybe "information-related" risk?  IDK

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,691 @@
+2. Information Hazards
+------
+
+Because the training data given to LLMs is so vast, there are cases (some involving litigation) of private information such as home addresses being leaked by LLMs.  Additionally, LLMs might correctly infer and disclose sensitive information while amassing and summarizing data (for example, accurately disclosing a serious disease diagnosis to an unprepared user asking about their symptoms).
```
```suggestion
Because the training data given to LLMs is so vast, there are cases of private information such as home addresses being leaked by LLMs.  Additionally, LLMs might correctly infer and disclose sensitive information while amassing and summarizing data (for example, accurately disclosing a serious disease diagnosis to an unprepared user asking about their symptoms).
```

franzenr replied:
Ah, I totally get wanting to highlight legal risks -- I think for me, this quick paranthetical didn't read as enough focus to actually be highlighting legal risks, so it felt more to me just like additional, but overall unnecessary, information about the fact that this is something that has already been observed. 

If we want to highlight legal risks, which seems valuable, I'm thinking it might make sense to more explicitly highlight them? Also, the legal risk here in particular seems to be more a risk to the LLM developers, rather than to the LLM end users, to whom this module is aimed. 

I might be misunderstanding the kinds of cases that have been brought forward, though! If there are cases that are being brought against users for giving someone else's sensitive data to a conversational agent, and then that data winding up in the training data, and then subsequently the model, then I think we should _definitely_ very explicitly mention that as a legal risk, maybe even in its own section, or at least small para?

pm0kjp replied:
I think you're on to something with the legal risk being to LLM developers here... I'll omit!

franzenr reviewed (changes requested):
first round so you have something to start chewing on!

pm0kjp commented:
@franzenr I addressed most of your issues, pushed back on a few, and also abbreviated the long example texts generated by AI, which I *thought* you pointed out as a flaw but now I can't find the comment?  LMK if this adequately meets your expectations for a PR!

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,691 @@
+@end
+
+@version_history 
+No previous versions.
```
Oh hmm, okay!!! I haven't been following the development of this particular template shift very closely, so I was just going by the QA guidelines -- glad one of us is in the know!

franzenr replied:
Yep, Rose H chimed in and that is true (And she has a PR out now to update docs about it) 

So DEFINITELY ignore this suggestion!

franzenr commented:
I definitely remember mentioning that to you at some point, but also don't know exactly where I did it!!!! 

I've given more context or clarification to the places where you pushed back or left things unresolved, so definitely take a look there when you have a chance. 

I'm going to keep going through things and will hopefully send through another review soon

pm0kjp commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,630 @@
+
+Because the training data given to LLMs is so vast, there are cases (some involving litigation) of private information such as home addresses being leaked by LLMs.  Additionally, LLMs might correctly infer and disclose sensitive information while amassing and summarizing data (for example, accurately disclosing a serious disease diagnosis to an unprepared user asking about their symptoms).
+
+While the authors do not explicitly mention this, another information hazard we want to point out is the risk related to input text.  Users can disclose private or regulated information to an LLM via prompts.  One of the ways that LLMs acquire additional language samples is by adding the prompts that users supply to their vast collection of language examples.  Supplying private information (about ...
```
```suggestion

<div class = "important">
<b style="color: rgb(var(--color-highlight));">Important note</b><br>

While the authors do not explicitly mention this, another hazard we want to point out is the risk related to input text.  Users can disclose private or regulated information to an LLM via prompts.  One of the ways that LLMs acquire additional language samples is by adding the prompts that users supply to their vast collection of language examples.  Supplying private information (about human subjects, about companies, and so on) when asking LLMs to generate text could constitute an impermissible breach of privacy or violation of intellectual property law or other regulation.  This risk isn't technically an information hazard, in the traditional sense, but it is information-adjacent.  

I</div>
```

pm0kjp replied:
@franzenr what do you think?

franzenr replied:
I think this is important enough to be a warning highlight box, not an "important note" box. I also think its important enough that the language should be more explicit about "HEY DO NOT DO THIS" as in my suggestion. 

If you prefer the rest of the language of your version, that's fine, but I think this is crucial enough that we should be highlighting the actionable takeaway from this much more explicitly. (This seems significantly more important to me to highlight than, e.g., not giving private info to regex101, which of course is still important, but has much lower risk in terms of info leakage.) 

If you don't think this is the place for that to be highlighted, I'm also okay with that -- just suggest another place to make this a warning. I'm not sure I'm comfortable passing this through QA without a warning notice somewhere that says in plain language what not to do. 

@pm0kjp

pm0kjp commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,630 @@
+>    - Detailed description of any statistical methods used.<br/>
+> <br/>
+> Note: The actual content and structure of the manuscript will depend on the specific details of the clinical trial, research area, and journal guidelines.<br/>
+Summarizing 
```
```suggestion

Summarizing 
```

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,630 @@
+@end
+
+@version_history 
+ 
```
```suggestion
 No previous versions.
```

Seems my erroneous suggestion to leave this blank per the docs was committed -- Rose H clarified that yes, this should NOT be left blank, and updated [docs](https://github.com/arcus/education_modules/blob/main/docs.md) accordingly.

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,630 @@
+@version_history 
+ 
+@end
+
```
```suggestion

good_first_module: false 
```

Feel free to change false to true if you want -- I don't really have strong opinions about how we're determining what a good first module is. From docs: "If this was a learner's very first experience with DART, would you be happy with this module being their first impression?" 

Required for all modules."

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,630 @@
+ 
+@end
+
+import: https://raw.githubusercontent.com/arcus/education_modules/main/_module_templates/macros.md
```
```suggestion
@sets_you_up_for

@end

@depends_on_knowledge_available_in

@end 

import: https://raw.githubusercontent.com/arcus/education_modules/main/_module_templates/macros.md
```


it was recently clarified that these are required fields for all modules (in addition to the good first module above). these two are perfectly fine to be left blank.

franzenr reviewed (changes requested):
I will have some more things to push through (mostly just small tweaks to alt text language, etc.) but I'm running in to my speaker coaching session now.

pm0kjp commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -286,49 +288,7 @@ Prompt: **"Make an outline for a peer-reviewed manuscript describing a clinical
->    - Recognition of individuals or organizations that contributed significantly to the study.<br/>
->    - Declaration of any conflicts of interest.<br/>
-> <br/>
+> ...
```
fixed in new commit!

pm0kjp commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,630 @@
+
+</div>
+
+## Quiz
```
fixed in new commit!

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,645 @@
+
+* [The manuscript describing ELIZA](https://dl.acm.org/doi/10.1145/365153.365168)
+
+Drawing Things:
+
+* [academo's 3D Vector Plotter](https://academo.org/demos/3d-vector-plotter/)
+* [NN-SVG](http://alexlenail.me/NN-SVG/index.html)
+
```
```suggestion
```
Not really related to the module topic, and if people really are interested in playing with the drawing stuff, you've already linked to these when you reference the sources within module content.

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,669 @@
+
+To understand neural networks, it's important to know that in machine learning (whether that's in a neural network or in another type of algorithm), there's a lot of trial and error.  In machine learning, the speed of calculation that a computer can offer allows the computer to try thousands of different values for the coefficient of a predictor variable, for example.  This allows the computer to ...
+
+![A figure with a hat states, "Oh, hey, you organized our photo archive!".  A second figure replies, "Yeah, I trained a neural net to sort the unlabeled photos into categories."  The first figure resonds, "Whoa! Nice work!."  The caption below the drawing reads: "Engineering tip: When you do a task by hand, you can technically say you trained a neural net to do it."](https://imgs.xkcd.com/comics/ ...
```
```suggestion
![A one-panel cartoon. A figure with a hat is looking at a phone and states, "Oh, hey, you organized our photo archive!".  A second figure replies, "Yeah, I trained a neural net to sort the unlabeled photos into categories."  The first figure responds, "Whoa! Nice work!"  The caption below the drawing reads: "Engineering tip: When you do a task by hand, you can technically say you trained a neural net to do it."](https://imgs.xkcd.com/comics/trained_a_neural_net.png "[Trained a Neural Net](https://xkcd.com/2173/) comic by xkcd, [CC BY-NC 2.5](https://xkcd.com/license.html).")<!-- style = "max-width: 220px;" -->Neural networks take this power of trial and error and massively expand it. Neural networks are an attempt to in some way replicate the operations of neurons in a brain. Various nodes (neurons) receive input, perform some mathematical transformation on that input, and either "fire" (pass some output on to a next step) or not, depending on whether the result of the calculation passes a certain threshold.  Each node starts off with random weights that determine how to mathematically transform the data it receives, and the computer tries many combinations of weights to see which ones are the most accurate. 
```

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,669 @@
+
+Let's consider a binary question.  Will a given person buy takeout for dinner?  Given seven input variables, you want a single outcome: 1 for "yes, they will get takeout" and 0 for "no, they will not get takeout."
+
+In this image (created thanks to [NN-SVG](http://alexlenail.me/NN-SVG/index.html)), the leftmost column of circles represent **inputs**.  Maybe one input is an individual's level of tiredness, another is the number of days they have had takeout in the last week, a third input could be the number of leftovers in the person's fridge.  These inputs are scaled and normalized.
```
```suggestion
In following image (created thanks to [NN-SVG](http://alexlenail.me/NN-SVG/index.html)), the leftmost column of circles represent **inputs**.  Maybe one input is an individual's level of tiredness, another is the number of days they have had takeout in the last week, a third input could be the number of leftovers in the person's fridge.  These inputs are scaled and normalized.
```

just as a cue for non-sighted learners that they haven't missed anything yet, since we're already launching into contextualization here.

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,669 @@
+
+In this image (created thanks to [NN-SVG](http://alexlenail.me/NN-SVG/index.html)), the leftmost column of circles represent **inputs**.  Maybe one input is an individual's level of tiredness, another is the number of days they have had takeout in the last week, a third input could be the number of leftovers in the person's fridge.  These inputs are scaled and normalized.
+
+![A depiction of a neural net, which consists of three columns of circles, the first of which has seven circles and is the input layer, the second of which has ten circles and is the hidden layer, and the last of which has one circle and is the output layer.  Lines connect the circles from the first column to circles in the second column, and circles in the second column to the circle in the thir ...
```
```suggestion
![A depiction of a neural net, which consists of three columns of circles, representing the input, hidden, and output layers. The input layer consists of seven circles, the hidden layer has ten circles, and the output layer has just one circle. Each circle of the input layer is connected by a line to each of the circles in the hidden layer, creating a large, overlapping web of connections. The lines are of varying darkness, with some so faint they are almost invisible. Then, each circle in the hidden layer has just one line connecting it to the output layer's single circle. These lines are also of varying darkness.](media/example_nn.png)Lines of various intensity show the **weights** (think of them like coefficients in an equation) that each of these inputs contribute to ten nodes in the **hidden layer**, which is a set of nodes between input and output. The darker the line, the heavier that input is weighted for that node's calculation. Some nodes might heavily consider one's previous takeout habits, for example, while others disregard it entirely. Each hidden layer node calculates the sum of all its weighted inputs, and each one comes up with its own unique answer.<br/><br/>
```

added some more contextualizing info about the visual characteristics of the very many lines between columns 1 & 2, and clean single line between 2 & 3, which I think is an important part of the visual representation here.

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,669 @@
+<div class = "important">
+<b style="color: rgb(var(--color-highlight));">Important note</b><br>
+
+![A figure with a ponytail indicates some words projected onto a screen while saying "An analysis of our new AI hiring algorithm has raised some concerns."  On the screen is displayed "DeepAIHire Candidate Evaluation Algorithm, Inferred Internal Weightings.  What follows is a table of weights that each apply to various factors.  There are four tiny weights and one enormous weight.  They are: weig ...
```
```suggestion
![A one-panel cartoon. A person with a ponytail indicates some words projected onto a screen while saying "An analysis of our new AI hiring algorithm has raised some concerns."  On the screen is displayed "DeepAIHire Candidate Evaluation Algorithm, Inferred Internal Weightings. Below that is a table of various factors and the weights applied to them. There are four tiny weights of values less than 1, and one enormous weight of more than 750.  The factors with tiny weights are Educational Background, Past Experience, Recommendations,  and Interview Performance. The factor with a massive weight is listed as "Enthusiasm for developing and expanding the use of the DeepAIHire Algorithm."](https://imgs.xkcd.com/comics/ai_hiring_algorithm_2x.png "[AI Hiring Algorithm](https://xkcd.com/2237/) comic by xkcd, [CC BY-NC 2.5](https://xkcd.com/license.html).")<!-- style = "max-width: 280px;" -->This risk framework reflects a point in time, and you may think of risk types or examples that aren't reflected here. Suffice it to say, LLMs, like any laboratory tool, must be used with discretion and an awareness of the risks, costs, tradeoffs, and utility they provide.
```
specific values aren't particularly important here -- I think just framing the discrepancy between the values at the start but not giving precise values, and then running through the lists of the factors given small or huge weights makes it easier to parse, and also does a better job preserving the humor.

franzenr commented on demystifying_large_language_models/demystifying_large_language_models.md:
```diff
@@ -0,0 +1,664 @@
+
+[[ ]] A tool used for translating languages on the internet.
+[[ ]] A type of supercomputer used for complex data analysis.
+[[ ]] A neural network-based AI model capable of understanding and generating human-like text.
```
```suggestion
[[X]] A neural network-based AI model capable of understanding and generating human-like text.
```

answer needs to be marked correct

pm0kjp replied:
WHOOPS!  TY!

franzenr commented:
I think the use of highlight boxes isn't entirely consistent with our standards, or at least feels a bit overused to the point where "important note" doesn't seem that notable anymore, BUT I also think could be subjective enough of an opinion, and would take long enough to change, that I'm ok just pushing this through for now

franzenr reviewed (approved):
LGTM
